dark-mode: Add a proposal for dark mode extension icons (#585)
hober/web-extensions: proposals/dark_mode_extension_icons.md

# Dark Mode Extension Icons

**Summary**

Feature to enable developers to enhance extension icon visibility in dark mode.

**Document Metadata**

**Author:** &lt;solomonkinard>

**Sponsoring Browser:** Chromium.

**Contributors:** &lt;oliverdunk>

**Created:** 2024-04-05

> **Related Issues:** &lt;https://crbug.com/893175>

Reviewer: The link is broken because of `>` (you should use the markdown link format). Also, I suggest adding the WECG link #229 here.

Author: Addressed in the patch response.

## Motivation

### Objective

Extension developers will be able to supply and define a set of icons to be used
in the event that browser system settings are set to dark mode. The set of dark
mode icons can and will be made automatically visible once the system setting is
in dark mode, thereby enhancing the overall browser experience.

#### Use Cases

1. Improved icon visibility in the extension toolbar.

> 2. Improved icon visibility on management the management and shortcuts pages.

Reviewer: Suggest adding following use cases:
- Improved icon visibility on context menu (Chrome and Safari reply on default icons, only Firefox can specify icons)
- Improved icon visibility on side panel (Chrome replies on default icons, Firefox  can specify icons)

Author: Addressed in the patch response.

1. Dark mode icon declarations made possible through the extension manifest.

> 3. setIcon() will allow storage of dark mode specific icons.

Reviewer 2: "storage"? Do you mean "setting"? IIRC, the icons set by this API are not persistent.

Author: Addressed in the patch response.

### Known Consumers

The associated and linked bug has 48 stars and 31 comments.

## Specification

### Schema

manifest.json
```
{
  "icon_variants": {
    "dark": {
      "128": "icons/128_dark.png"
    },
    "light": {
      "128": "icons/128_light.png"
    }
  },
  "action": {
    "default_icon": {
      "128": "icons/128.png"
    },
    "icon_variants": {
      "dark": {
        "128": "icons/128_action_dark.png"
      },
      "light": {
        "128": "icons/128_action_light.png"
      }
    }
  }

> }

Reviewer 2: Like `default_icon` in `action`, the `light` and `dark` values should support a string path, and not just object values, to support single SVG icon files.

Like this:

```json
{
  "icon_variants": {
    "dark": "icons/dark_icon.svg",
    "light": "icons/light_icon.svg"
  },
  "action": {
    "default_icon": "icons/default_icon.svg",
    "icon_variants": {
      "dark": "icons/action_dark_icon.svg",
      "light": "icons/action_light_icon.svg"
    }
  }
}
```

Can you include that in this proposal? Safari (and likely Firefox) would implement this.

Author: Addressed in the patch response.

Reviewer 3: Hmm, I don't think this is addressed? See my other comment :)

Author: I changed to idl to standardize the format. I think that it was addressed in the latest patch by adding a string option anywhere icon_variants is used.

```

Browser action.
```
action.setIcon({
  imageData?: ImageData | object,
  path?: string | object,
  tabId?: number,
  mode?: string

> });

Reviewer 2: I wonder if developers will expect the `object` values for `imageData` and `path` to take the new `light` and `dark` values inside them, instead of using the `mode` property. (That's kind of what I would expect.) That would also allow setting both modes at the same time instead of requiring two calls to `setIcon` for each mode.

I guess `imageData` and `path` can't take the new format for compatibility concerns, but I wonder if new properties that do take the new object format should be added instead of `mode`? `imageDataVariants` and `pathVariants` perhaps?

I think this part of the proposal needs some more thought.

Author: Addressed in the patch response.

```

### Behavior

Existing manifest key. The behavior of the icons manifest key will remain unchanged.
```
{
  "icons": {
    "64": "icon_64.png"
  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

### New Permissions

N/A.

### Manifest File Changes

A new icon_variants key and sub-key will be added. The schema is listed in the
schema section. Failure to parse this key and/or subkey will prevent the
extension from loading. `icon_variants` are required in the manifest and there
will be no warning whether they're present or not. If one or both
`icon_variants` keys are present, they must include at least a "dark" or "light"
sub-key. Failure to do so will result in an error. Any of those theme keys must
have at least one size key as a string, containing the path to the icon as the
value. Failure to find any file at the specified path will result in an error.

> Such errors prevent the extension frm loading.

Reviewer 2: This paragraph is currently pretty dense and hard to follow. Maybe break it up into a few paragraphs?

Is `icon_variants` required, or is optional? What does "required in the manifest and there will be no warning whether they're present or not" mean? That sounds like optional.

How does fallback happen with `icons` and `default_icon` if `icon_variants` only has `light` or `dark`? That would be good to mention here. (It is sort of implied in the "precedence" sentence above.)

The word "theme" is include only once in this document, everywhere else "mode" is used. Perhaps change "theme" to "mode" here?

In general, I think blocking the extension from loading in case of a recoverable manifest errors is the wrong choice, and it is one of the reasons we couldn't reuse the `icons` key for this feature. I support the browser warning about bad manifest syntax or types, but having this be a hard failure seems too strong. Especially when fallback to `icons` or `default_icon` is an option. This will at least not be a hard error in Safari, so maybe leave this up to the browser?

As mentioned before, I think allowing string values for single file SVG support should be allowed, and that should be mentioned here too.

Typo on the last line ("frm").

Author: Addressed in the patch response.

1. A developer could ask the user to specify what mode they're in and then
dynamically set the icon to a dark mode icon using action.setIcon(). That
wouldn't change the management page icon.
2. A developer could change icons dynamically to dark mode icons if this is
true: `window.matchMedia('(prefers-color-scheme: dark)')`.

### Open Web API

As stated in the workarounds section, the following is already an option to consider: `window.matchMedia('(prefers-color-scheme: dark)')`. However, that
wouldn't automatically set icons dynamically, as it would require subsequent
calls to action.setIcon(). It also wouldn't update the icon on the management
pages.

## Implementation Notes

N/A.

## Future Work

#comment13 in the linked bug asks for the action popup background color to be
changed [to become become dark] along with the icon in dark mode. That is out of
scope for this proposal as it would require rethinking the blank page in a
normal browser tab, in addition to what happens when loading an iframe on top of
the dark background. Also, which color should be chosen for the background.
Perhaps to seed the idea, one of these colors could be analyzed and compared to

> others not included here: #202124, #2D2E2F.

Reviewer 2: [Safari supports dark mode in popups](https://developer.apple.com/documentation/safariservices/safari_web_extensions/optimizing_your_web_extension_for_safari?language=objc) by detecting the `color-scheme` CSS property. I suggest that be adopted as the preferred technique for action popups.

Reviewer: I think @solomonkinard and @xeenon have some misunderstandings. Yes, this proposal only solves the standalone icon problem. The light and dark themes of an extension page (popup page) are not in scope, there are standard Web solution for it (and it is already supported by browsers).

So I suggest removing this paragraph.

Reviewer 2: Yes the popup is out of scope for this document and should just be dropped to avoid confusion or scope creep.

Author: Addressed in the patch response.

# Dark Mode Extension Icons

**Summary**

Feature to enable developers to enhance extension icon visibility in dark mode.

**Document Metadata**

**Author:** solomonkinard

**Sponsoring Browser:** Chromium, Safari, Firefox

**Contributors:** oliverdunk, xeenon, carlosjeurissen

**Created:** 2024-04-05

**Related Issues:**
* https://crbug.com/893175
* https://github.com/w3c/webextensions/issues/229

## Motivation

### Objective

Extension developers will be able to supply and define a set of icons to be used
in the event that user has expressed the preference for a page that has a dark

> theme. The set of dark mode icons can and will be made automatically visible

Reviewer 3: Thoughts on dropping the second line here? We could also keep it, but if we do that I think we should remove the mention of a "system setting" to avoid dictating what the browser uses when determining the preferred mode.

Author: Done.

once the system setting is in dark mode, thereby enhancing the overall browser
experience.

#### Use Cases

1. Improved icon visibility in the extension toolbar.

> 1. Improved icon visibility on management the management and shortcuts pages.

Reviewer 3: ```suggestion
1. Improved icon visibility on the management and shortcuts pages.
```

1. Dark mode icon declarations made possible through the extension manifest.
1. setIcon() will allow setting of dark and/or light mode specific icons.

> 1. Improved icon visibility on context menu (Chrome and Safari reply on default

Reviewer 3: ```suggestion
1. Improved icon visibility on context menu (Chrome and Safari rely on default
```

icons, only Firefox can specify icons).

> 1. Improved icon visibility on side panel (Chrome replies on default icons,

Reviewer 3: ```suggestion
1. Improved icon visibility on side panel (Chrome relies on default icons,
```

Firefox can specify icons).

### Known Consumers

The Chromium bug has a significant amount of developer interest.

## Specification

### Schema

manifest.json
```
"icon_variants": [
  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png",
  },
  {

>     "16": "dark16.png",

Reviewer 3: A remaining ambiguity is how the browser knows what is a special property and what is a size. Presumably we should just say something like "keys made entirely of numbers are reserved, cannot be used as properties and will always be treated as a size"? cc/ @xeenon for thoughts

Reviewer: size keys:
- number string: "16", "32", "48", "96", "128", same as `icons`.
- "any": is reserved for svg.

non-size keys:
- "color_scheme"

All other keys are unknown or unsupported at the moment.

Reviewer 3: I'm not sure if this is actually true? My understanding (which I haven't tested much, but seems to match the implementation and docs) is that we will accept any size in the icons key, there are just some that are preferred. So if we really want to be consistent I think we need to allow any size.

Reviewer 2: @oliverdunk Correct, any number should be a valid size (it is for `icons` in Safari). Every other property is not a size (besides `any`). Having a fixed list of sizes should not be specified, as display densities and browser differences necessitate different sizes. (Safari has different preferred sizes when compared to Chrome.)

@hanguokai I would not specify `any` as "reserved for SVG". It could be used by any file type — maybe it is a 512px PNG and the developer just wants the browser to scale it as needed. That should be supported. This is important for platforms like visionOS, where the user can walk right up to an icon that fills their field of view, and the dynamic rendering might need an icon that large. While ideally that would be an SVG, a high pixel density bitmap should be fully supported too.

Author: Done. Added:
```1. 16 is a size in `{"16": "icon.png"}` and any number as a size can be
used, as per browser specifications. The word `"any"` could also be used in
place of a number to represent a path to an icon with any supported format.```

    "32": "dark32.png",
    "color_scheme": "dark",
  },
  {
    "16": "light16.png",
    "32": "light32.png",
    "color_scheme": ["dark", "light"]
  }
]
```

Set icon_variants dynamically.
```
const exampleProperties: {string: string | ImageData}[] = [
  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png"
  },
  {
    "16": darkImageData16,
    "32": darkImageData32,
    "color_scheme": "dark"
  },
  {
    "16": lightImageData16,
    "32": lightImageData32,
    "color_scheme": "light"
  }
];

// action.setIcon().

> const actionProperties = {variants: exampleProperties};

Reviewer 3: ```suggestion
const createProperties = {variants: exampleProperties};
```

action.setIcon(createProperties);

// menus.*(), for supporting browsers.

> const menusProperties = {icon_variants: exampleProperties}

Reviewer 3: ```suggestion
const menusProperties = {icon_variants: exampleProperties};
```

menus.create(menusProperties);
menus.update(menusProperties);
```

A benefit of this new structure is that it's more resiliant to future changes, thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
monochrome), and etc.

## setIcon()
Incumbent action.setIcon(), for reference.
```
action.setIcon({
  path,
  tabId
  imageData,
});
```

### Behavior

Existing manifest key. The behavior of the icons manifest key will remain unchanged.
```
{
  "icons": {
    "64": "icon_64.png"
  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

The `dark` icon variant is used when a dark color scheme is used for the
surrounding UI, often referred to as "dark mode".
The `light` icon variant is used otherwise.

### New Permissions

N/A.

### Manifest File Changes

1. If icon_variants are supplied, icons can be ignored.

> 1. If icon_variants contain an icon object duplicate, the first match will be

Reviewer 3: ```suggestion
1. If icon_variants contain an icon object with duplicate conditions, the first match will be
```

used.
1. Any icon object that does not contain a "color_scheme" key will apply to both
light and dark.

> 1. If only one icon object is supplied, it will be used for both light and dark.

Reviewer 3: Even if `color_scheme: "dark"` is specified? I would lean towards saying that if the browser recognises the icon_variants key, it can error if there is not at least one light mode (or light and dark mode etc.) icon provided.

Reviewer: I'm leaning toward not a hard error, but could be a warning.

Reviewer 4: @oliverdunk having a check if at least one light mode icon is specified would be helpful. If we make it a warning instead of an error we would need to figure out what to do instead. @hanguokai do you believe the browser should simply use the dark icon instead?

In general it seems a hard error would motivate developers to configure the `icon_variants` correctly. I do not see backwards compatibility issues for hard errors in this situation.

Reviewer: @carlosjeurissen For now, browsers are quite tolerant of `icons`. For example, if developers don't provide `icons` or `icons` is an empty array (no one icon), the browser just ignore it and display a default icon without any error or warning. Of course, developers generally do not do this. However, what is discussed here is how browsers handle the most extreme situations. I think tolerating is better than reporting errors, just like now. And in my option, if only one icon is provided, the browser no longer needs to match any conditions. This is also a bit like the relationship between programming language engine and linter. The engine only needs to make sure it works properly, and the extra linter tool (e.g. a IDE addon) makes sure your code best.

> 1. icon_variants will not cause an error in the event that it is invalid

Reviewer 3: Can we make this more specific? With the current wording, not recognising a single key in a single icon object would allow us to ignore the entire icon_variants key. I think we need to say that we will skip over individual icons in the array that contain a key we don't recognise.

Author: Done.

Instead, it can be ignored altogether or just emit a warning. Warnings are
preferred over errors because they're more adaptable to changes in the future.
1. Neither "dark" nor "light" color_scheme's are required.

> 1. Icon objects missing the color_scheme property will apply to light and dark

Reviewer 3: I think this is a duplicate of "Any icon object that does not contain a "color_scheme" key will apply to both
light and dark"?

Author: Removed.

  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png"
  },
  {
    "16": darkImageData16,
    "32": darkImageData32,
    "color_scheme": "dark"
  },
  {
    "16": lightImageData16,
    "32": lightImageData32,
    "color_scheme": "light"
  }
];

// action.setIcon().
const createProperties = {variants: exampleProperties};
action.setIcon(createProperties);

// menus.*(), for supporting browsers.
const menusProperties = {icon_variants: exampleProperties};
menus.create(menusProperties);
menus.update(menusProperties);
```

A benefit of this new structure is that it's more resiliant to future changes, thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
monochrome), and etc.

## setIcon()
Incumbent action.setIcon(), for reference.
```
action.setIcon({
  path,
  tabId
  imageData,
});
```

### Behavior

Existing manifest key. The behavior of the icons manifest key will remain unchanged.
```
{
  "icons": {
    "64": "icon_64.png"
  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

> The `dark` icon variant is used when a dark color scheme is used for the
> surrounding UI, often referred to as "dark mode".

Reviewer 4: Considering some browsers, while in light mode, might want to use the dark icon variant for better readability if a darker theme is used. I would suggest to rephrase it to something like this:

```suggestion
The `dark` icon variant can be used by browsers to improve readability of the icon. Often this will be when the os/browser color scheme is dark (often referred to as "dark mode") but also when the browser uses a darker shade theme which has a background on which the dark icon variant would be more readable.
```

Reviewer 3: I agree with the general direction, but could we say something like "situations where the browser may choose to use this include", and then have both as bullet points? I'd like to avoid anything that gives the impression it is dictating when the browser picks them.

Reviewer 4: @oliverdunk Sure! In general that would make it more readable as well. In any case the intention of the suggestion was to give potential examples. The whole purpose of using `color_scheme` and the syntax we are after is to go for the best contract with the intentions of the developer with the decisions made by the browser.

The `light` icon variant is used otherwise.

### New Permissions

N/A.

### Manifest File Changes

1. If icon_variants are supplied, icons can be ignored.
1. If icon_variants contain an icon object with duplicate conditions, the first match will be
used.
1. Any icon object that does not contain a "color_scheme" key will apply to both
light and dark.
1. If only one icon object is supplied, it will be used for both light and dark.
1. icon_variants will not cause an error in the event that it is invalid
Instead, it can be ignored altogether or just emit a warning. Warnings are
preferred over errors because they're more adaptable to changes in the future.
1. Neither "dark" nor "light" color_scheme's are required.
1. Icon objects missing the color_scheme property will apply to light and dark
mode. It could also be explicity set using {"color_scheme": ["dark", "light"]}.
1. If icon_variants are provided, the top level icons key and

> action.default_icon will be ignored. There is no fallback.

Reviewer 4: Considering the `icon_variants` being present. `action.default_icon` being set and `action.icon_variants` missing.  As follows:
```json
{
  "icon_variants": [...],
  "action": {
    "default_icon": "...path"
  }
}
```

Intuitively it feels `action.default_icon` should override whatever is set in the main `icon_variants`. Same for setting [`action.theme_icons` ](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/action) would override what is set in the main `icon_variants`.

My proposal would be:
```suggestion
If the top level icon_variants key is provided, the top level icons key will be ignored.
For action icons follows a strict order of precedence:
action.icon_variants > action.theme_icons > action.default_icon > icon_variants > icons
Any programmatically set icons using `action.setIcon` would override the declaratively defined icons mentioned above.
```

Reviewer 3: I like this.

1. 16 is a size in `{"16": "icon.png"}` and any number as a size can be
used, as per browser specifications. The word `"any"` could also be used in
place of a number to represent a path to an icon with any supported format.

> 

Reviewer 4: To make user extension developers can rely on deterministic behaviour I suggest we specify how icon size fallbacks will be handled by browsers:
```suggestion
The icon size used by the browser will be determined as follows:
For raster images. The requested size will be used as defined in icon_variants. If the requested icon size by the browser is not present in the manifest file, a size higher than the requested icon size will be used if present.
For vector images. The size lookup will be density independent.
Example: the browser wants to use an icon with size 16 with a density 1.5. For raster images, it will search for an icon with size 24. While for vector images it will search for an icon with size 16.
```
This specific behaviour for vector images allows extensions to better control the amount of details they want to display in icons.

Reviewer 3: I feel like this might be over-specifying, but interested to hear from other browser vendors.

Reviewer 2: I think it is worth writing this down, even though it mostly follows the old rules. We should specify the unspecified from before. Plus enough has changed with the addition of `any` and SVG.

Reviewer 4: @xeenon Agreed. We should probably also specify what should happen when both explicitly size (16, 32) and "any" are defined in the manifest. Should this be valid? I see a situation in which a browser would use the exact size if present, otherwise fall back to "any". How do you feel about that?

When it comes to vector size matching. If we want to go all in there are a few modes on which matching could take place.
1) DENSITY closest size * density. This is how matching would work for raster images with the exception that it is not rounded up as a closest match would not lead to side effects (blurry images).
2) EXACT closest size density independent.
3) Most specific dimensions based on integer ratios. This one is a bit more complicated. To reach pixel perfect icons, authors should specify the viewBox dimensions as size. The browser will then attempt to find the closest image size resulting in crispy images. Example: The browser wants a 48px icon. If the extension defines both 32px and 24px sizes in the manifest. The browser will pick 24px as this will make sure the icon is as pixel perfect as possible.

The mode used could potentially be explicitly set in the icon variants image group. Something like:
```json
[{
  "24": "path",
  "32": "path",
  "sizeMatching": ["ratio", "exact"]
}]
```

Reviewer: At present, there is no [extensions specification](https://w3c.github.io/webextensions/specification/) for `icons` at all. But browsers are already very good at this. For example, there is "[preferred density-corrected dimensions](https://html.spec.whatwg.org/multipage/images.html#preferred-density-corrected-dimensions)" in the image specification. Another example is `<source>` elements in a `<picture>` element.

The icons in the extension is simpler than images in web pages. For now, we could leave icon-matching to the browser implementation. In the future, if necessary, we can gradually improve the specification.

Reviewer 2: @carlosjeurissen I think 3 makes the most sense, while 1 is what we do today. I would likely do 3 when implementing this for Safari / WebKit.

Reviewer 4: @xeenon mode 3 has the downside of making it less intuitive / easy for authors to add / remove details based on the sizes being used. One may also wonder how a potential density property would play a role here.

# Dark Mode Extension Icons

**Summary**

Feature to enable developers to enhance extension icon visibility in dark mode.

**Document Metadata**

**Author:** &lt;solomonkinard>

**Sponsoring Browser:** Chromium.

**Contributors:** &lt;oliverdunk>

**Created:** 2024-04-05

**Related Issues:**
* https://crbug.com/893175
* https://github.com/w3c/webextensions/issues/229

## Motivation

### Objective

Extension developers will be able to supply and define a set of icons to be used

> in the event that browser system settings are set to dark mode. The set of dark

Reviewer 3: I'm not sure that "browser system settings" is really a meaningful term. The [prefers-color-scheme spec](https://drafts.csswg.org/mediaqueries-5/#prefers-color-scheme) says dark is used if the "user has expressed the preference for a page that has a dark theme"? To avoid over-specifying here, should we say something similar?

Author: Done.

mode icons can and will be made automatically visible once the system setting is
in dark mode, thereby enhancing the overall browser experience.

#### Use Cases

1. Improved icon visibility in the extension toolbar.
1. Improved icon visibility on management the management and shortcuts pages.
1. Dark mode icon declarations made possible through the extension manifest.
1. setIcon() will allow setting of dark and/or light mode specific icons.
1. Improved icon visibility on context menu (Chrome and Safari reply on default icons, only Firefox can specify icons).
1. Improved icon visibility on side panel (Chrome replies on default icons, Firefox can specify icons).

### Known Consumers

> The associated and linked bug has 48 stars and 31 comments.

Reviewer 3: Really a matter of preference, but I would just say something like "the Chromium bug has a significant amount of developer interest" since this is going to become out of date quickly.

Author: Done.

## Specification

### Schema

Some browsers don't currently have plans to support svg. Therefore, a drop-in
replacement of an already supported image type will be allowed.

numberToPathDictionary
```
{
  "<number>": "<path>"
}
```

manifest.json
```
{
  "icon_variants": {

>     "<dark|light>": <numberToPathDictionary>

Reviewer 3: I think to address @xeenon's feedback we should also accept a single icon here?

Author: Done.

  },
  "action": {
    "default_icon": <numberToPathDictionary>|<svgPathString|pathString>,
    "icon_variants": {
      "<dark|light>": <numberToPathDictionary>
    }
  }
}
```

Action
```
action.setIcon({
  imageData?: ImageData | object:<numberToPathDictionary>,
  path?: string | object:<numberToPathDictionary>,
  tabId?: number,
  variants?: {
    "<dark|light>": <numberToPathDictionary>
  },

> });

Reviewer: If I understand correctly, `imageData` is not supported by `variants` in `action.setIcon()`, right? Do we consider providing APIs to detect current dark/light theme and a onChangeListener?

Reviewer 3: @solomonkinard, is this an implementation limitation or just an accidental omission?

Author: Should ImageData be supported here in icon_variants? Note: `variants` here was changed to `icon_variants` for consistency.

Reviewer: > Note: `variants` here was changed to `icon_variants` for consistency.

For JavaScript api `action.setIcon()`, the property name should be `iconVariants` (Camel Case), not `icon_variants`.

Reviewer 2: It also should just be `variants` since an icon prefix is redundant.

Author: Done.

    "dark": {
      "128": "128_dark.png"
    },
    "light": {
      "128": "128_light.png"
    }
  },
  "action": {
    "default_icon": {
      "128": "128_action.png"
    },
    "icon_variants": {
      "dark": {
        "128": "128_action_dark.png"
      },
      "light": {
        "128": "128_action_light.png"
      }
    }
  }
}
```

Action
```
action.setIcon({
  variants?: {
    "dark": {
      "128": "128_action_dark.png"
    },
    "light": {
      "128": "128_action_light.png"
    }
  },
});
```

### Behavior

Existing manifest key. The behavior of the icons manifest key will remain unchanged.
```
{
  "icons": {
    "64": "icon_64.png"
  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

### New Permissions

N/A.

### Manifest File Changes

#### Optional or required?
`icon_variants` is an optional key of the top level manifest dictionary.
`icon_variants` is also an optional sub-key in the action key dictionary.

#### Fallback
`icon_variants` will be used if they're supplied. Otherwise `default_icon` will

> be used if it's supplied. If neither are supplied, then `icons` will be used.

Reviewer: About order of precedence and fallback.

1. If I understand correctly, the `icon_variants` or `variants` in `action.setIcon()` must provide both dark and light icon if they occur, right? This can simplify the fallback problem.

1. Order of precedence for action:
`API: action.setIcon()` → `action.icon_variants` → `action.default_icon` → `icon_variants` → `icons`

1. Order of precedence for non-action:
`icon_variants` → `icons`

Reviewer 3: It would definitely be nice to have some more detail here. For example:

- What happens if you specify icon_variants.dark but not icon_variants.light, and also icons? Do we use the dark mode icon from icon_variants for light as well, or fallback to icons?
- What happens if you specify icon_variants.dark but not icon_variants.light, and you don't have icons? Do we use the dark mode icon from icon_variants or show an auto-generated icon?

Author: TBA (to be addressed) in upcoming patch. Can you clarify `Do we use the dark mode icon from icon_variants or show an auto-generated icon?`

Reviewer 3: > Can you clarify Do we use the dark mode icon from icon_variants or show an auto-generated icon?

In this case you only explicitly specified a dark mode icon, so we can either use that for light mode as well, or just fallback to the browser's default behavior (which in Chrome is a grey icon with the first letter of the extension name).

Author: The new patch adds the following to one of your other related questions:

> Another option could be to fall back to
the dark mode icon if in light mode and no light mode icon is supplied.

I suppose that the proposal shouldn't necessarily have options in it, so it's an opportunity for input.

#### Warning or error?
The new `icon_variants` key (or sub-key) will only emit a non-blocking warning

> rather than a hard error in the event of unexpected values or missing icons.

Reviewer 3: Just a nit.

```suggestion
rather than a hard error in the event of unexpected keys, values or missing icons.
```

Author: Rephrased this entire section, as recommended by Rob.

future changes. The `icons` key cannot be used for dark mode support because it
errors in cases when a warning would have otherwise sufficed. The warning is
ok for `icon_variants` based on the safety net outlined in the fallback section.

## Security and Privacy

### Exposed Sensitive Data

N/A.

### Abuse Mitigations

N/A.

### Additional Security Considerations

N/A.

## Alternatives

### Existing Workarounds

1. A developer could ask the user to specify what mode they're in and then
dynamically set the icon to a dark mode icon using action.setIcon(). That
wouldn't change the management page icon.
2. A developer could change icons dynamically to dark mode icons if this is
true: `window.matchMedia('(prefers-color-scheme: dark)')`.

### Open Web API

As stated in the workarounds section, the following is already an option to consider: `window.matchMedia('(prefers-color-scheme: dark)')`. However, that
wouldn't automatically set icons dynamically, as it would require subsequent
calls to action.setIcon(). It also wouldn't update the icon on the management
pages.

## Implementation Notes

N/A.

## Future Work

> Svg support may be considered at some point for some browsers.

Reviewer 3: I would probably drop this, it doesn't feel related to the proposal (as much as I would love it!).

Author: Done.

# Dark Mode Extension Icons

**Summary**

Feature to enable developers to enhance extension icon visibility in dark mode.

**Document Metadata**

**Author:** solomonkinard

**Sponsoring Browser:** Chromium, Safari, Firefox

**Contributors:** oliverdunk, xeenon, carlosjeurissen, hanguokai

**Created:** 2024-04-05

**Related Issues:**
* https://crbug.com/893175
* https://github.com/w3c/webextensions/issues/229

## Motivation

### Objective

Extension developers will be able to supply and define a set of icons to be used
in the event that user has expressed the preference for a page that has a dark
theme.

#### Use Cases

1. Improved icon visibility in the extension toolbar.
1. Improved icon visibility on the management and shortcuts pages.
1. Dark mode icon declarations made possible through the extension manifest.

> 1. setIcon() will allow setting of dark and/or light mode specific icons.

Reviewer 2: ```suggestion
1. `setIcon()` will allow setting of dark and/or light mode specific icons.
```

1. Improved icon visibility on context menu (Chrome and Safari rely on default

> icons, only Firefox can specify icons).

Reviewer 2: ```suggestion
1. Improved icon visibility on context menu (Chrome relies on default
icons, only Firefox and Safari can specify menu icons).
```

1. Improved icon visibility on side panel (Chrome relies on default icons,
Firefox can specify icons).

### Known Consumers

The Chromium bug has a significant amount of developer interest.

## Specification

### Schema

manifest.json
```
"icon_variants": [
  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png",
  },
  {
    "16": "dark16.png",
    "32": "dark32.png",
    "color_scheme": "dark",
  },
  {
    "16": "light16.png",
    "32": "light32.png",
    "color_scheme": ["dark", "light"]
  }
]
```

> 

Reviewer 4: ```suggestion
  `icon_variants` requires an array with a set of icon groups objects. Each icon group consists of a set of icon paths and icon group matching criteria.
  
  The icon paths are set using the size as key and/or the keyword "any".
  
  Optionally the `color_scheme` matching criterium which can either be an string or array of color schemes.
```

  },
  {
    "16": darkImageData16,
    "32": darkImageData32,
    "color_scheme": "dark"
  },
  {
    "16": lightImageData16,
    "32": lightImageData32,
    "color_scheme": "light"
  }
];

// action.setIcon().
const createProperties = {variants: exampleProperties};
action.setIcon(createProperties);

// menus.*(), for supporting browsers.
const menusProperties = {icon_variants: exampleProperties};
menus.create(menusProperties);
menus.update(menusProperties);
```

A benefit of this new structure is that it's more resiliant to future changes,
thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
monochrome), and etc.

## setIcon()
Incumbent action.setIcon(), for reference.
```
action.setIcon({
  path,
  tabId
  imageData,
});
```

### Behavior

Existing manifest key. The behavior of the icons manifest key will remain
unchanged.
```
{
  "icons": {
    "64": "icon_64.png"
  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

The `dark` icon variant can be used by browsers to improve readability of the

> icon. Often this will be when the os/browser color scheme is dark (often

Reviewer 2: ```suggestion
icon. Often this will be when the OS or browser color scheme is dark (often
```

referred to as "dark mode") but also when the browser uses a darker shade theme
which has a background on which the dark icon variant would be more readable.
The `light` icon variant is used otherwise.

### New Permissions

N/A.

### Manifest File Changes

> 1. Any icon object that does not contain a "color_scheme" key will apply to all

Reviewer 4: ```suggestion
1. Any icon group that does not contain a "color_scheme" key will apply to all
```

> available options, e.g. both "light" and "dark".

Reviewer 2: ```suggestion
1. Any icon object that does not contain a `"color_scheme"` key will apply to all
available options, e.g. both `"light"` and `"dark"`.
```

> 1. If icon_variants contain an icon object with matching conditions, the first
> match will be used.

Reviewer 4: ```suggestion
1. If "icon_variants" contains an icon group with matching conditions, the icon(s)
specified in the first matching icon group based on insertion order will be used.
The other icon groups after that match will be ignored.
```

>    * "\<size\>" will be used used instead of "any" in case in case there are

Reviewer 2: ```suggestion
   * "\<size\>" will be used used instead of `"any"` in case in case there are
```

>    * "\<size\>" will be used used instead of "any" in case in case there are
>    matching conditions.

Reviewer 4: ```suggestion
```
This is explained under 7 below. I suggest to restrict this paragraph to icon group matching.

>    * If there is more than one matching icon object, any without `color_scheme` will be applied to every possible color scheme. Therefore, a subsequent matching icon object with a color_scheme will not be used.

Reviewer 4: ```suggestion
```
This is already implied by the phrase above stating "the first match will be used."

>    * If there is more than one matching icon object, any without `color_scheme` will be applied to every possible color scheme. Therefore, a subsequent matching icon object with a color_scheme will not be used.

Reviewer 2: ```suggestion
   * If there is more than one matching icon object, any without `color_scheme` will be applied to every possible color scheme. Therefore, a subsequent matching icon object with a `color_scheme` will not be used.
```

   * Non-exact size matches will return an icon nearest the requested size, or the smallest in case of a tie.
   * Incorrectly typed `color-scheme`'s in manifest.json lead to undefined

>    behavior.

Reviewer 2: ```suggestion
   * Incorrectly typed `color-scheme` will be ignored, with an optional warning.
```

   * If e.g. only one icon object is defined with a specific color scheme, that
   icon object will be applied to all color schemes. It will be treated as the
   default.

> 1. If only one icon object is supplied, it will be used for both light and dark.

Reviewer 4: ```suggestion
1. If only one icon group is supplied, all matching conditions will be ignored.
```

1. icon_variants will not cause an error in the event that it is invalid

> Instead, only the fauly icon object(s) will be ignored, with an optional

Reviewer 4: ```suggestion
Instead, only the faulty icon group(s) will be ignored, with an optional
```

warning. Warnings are preferred over errors because they're more adaptable to
changes in the future.

> 1. Neither "dark" nor "light" color_scheme's are required.

Reviewer 2: ```suggestion
1. Neither `"dark"` nor `"light"` color schemes are required.
```

> 1. If the top level `icon_variants` key is provided, the top level `icons` key

Reviewer 2: ```suggestion
1. If the top-level `icon_variants` key is provided, the top level `icons` key
```

will be ignored.
   1. `action` icons follow a strict order of precedence:\

>    action.icon_variants > action.theme_icons > action.default_icon > icon_variants > icons.

Reviewer 2: I'm also not sure `theme_icons` should be mentioned here, as that is Firefox only?

```suggestion
   1. For `action` icons, the browser first checks for `icon_variants` in the `action`. If not specified, it falls back to `default_icon`, then to the top-level `icon_variants` or `icons`.
```

>    1. Any programmatically set icons using `action.setIcon` would override the declaratively defined icons mentioned above.

Reviewer 2: ```suggestion
   1. Any programmatically set icons using `action.setIcon()` would override the declaratively defined icons mentioned above.
```

1. 16 is a size in `{"16": "icon.png"}` and any number as a size can be
used, as per browser specifications. The word `"any"` could also be used in
place of a number to represent a path to an icon with any supported format.
The icon size used by the browser will be determined as follows:
   1. For raster images. The requested size will be used as defined in icon_variants.
If the requested icon size by the browser is not present in the manifest file,
a size higher than the requested icon size will be used if present.
   1. For vector images. The size lookup will be density independent.
Example: the browser wants to use an icon with size 16 with a density 1.5. For
raster images, it will search for an icon with size 24. While for vector images

> it will search for an icon with size 16.

Reviewer 2: This can be cleaned up and add further clarity.

```suggestion
1. The `"16"` is a size in `{"16": "icon.png"}` and any number can be used as a size, per
the browser’s icon requirements. The word `"any"` can also be used in place of a number
to represent an icon that can be shown at any size (usually vector images). The icon size
used by the browser will be determined as follows:
   1. **Raster images:** Sizes are in pixels. For high-density devices (2x, 3x, etc.), the
browser looks for double or triple the desired point size (e.g., 32 or 48 pixels for a 16
point request). If the exact pixel size is unavailable, the next largest pixel size or `"any"`
will be used.
   1. **Vector images:** Sizes are in points, ensuring device independence. If the exact
point size is unavailable, an integer multiple (e.g. 32, 48, etc.) or `"any"` will be used.
```

## Security and Privacy

### Exposed Sensitive Data

N/A.

### Abuse Mitigations

N/A.

### Additional Security Considerations

N/A.

## Alternatives

### Existing Workarounds

1. A developer could ask the user to specify what mode they're in and then

> dynamically set the icon to a dark mode icon using action.setIcon(). That

Reviewer 2: ```suggestion
dynamically set the icon to a dark mode icon using `action.setIcon()`. That
```

wouldn't change the management page icon.
2. A developer could change icons dynamically to dark mode icons if this is
true: `window.matchMedia('(prefers-color-scheme: dark)')`.

### Open Web API

As stated in the workarounds section, the following is already an option to
consider: `window.matchMedia('(prefers-color-scheme: dark)')`. However, that
wouldn't automatically set icons dynamically, as it would require subsequent

> calls to action.setIcon(). It also wouldn't update the icon on the management

Reviewer 2: ```suggestion
calls to `action.setIcon()`. It also wouldn't update the icon on the management
```

# Dark Mode Extension Icons

**Summary**

Feature to enable developers to enhance extension icon visibility in dark mode.

**Document Metadata**

**Author:** solomonkinard

**Sponsoring Browser:** Chromium, Safari, Firefox

**Contributors:** oliverdunk, xeenon, carlosjeurissen, hanguokai

**Created:** 2024-04-05

**Related Issues:**
* https://crbug.com/893175
* https://github.com/w3c/webextensions/issues/229

## Motivation

### Objective

Extension developers will be able to supply and define a set of icons to be used

> in the event that user has expressed the preference for a page that has a dark
> theme.

Reviewer 5: Missing a word.

```suggestion
in the event that the user has expressed the preference for a page that has a
dark theme.
```

#### Use Cases

1. Improved icon visibility in the extension toolbar.
1. Improved icon visibility on the management and shortcuts pages.
1. Dark mode icon declarations made possible through the extension manifest.
1. `setIcon()` will allow setting of dark and/or light mode specific icons.
1. Improved icon visibility on context menu (Chrome relies on default
icons, only Firefox and Safari can specify menu icons).
1. Improved icon visibility on side panel (Chrome relies on default icons,
Firefox can specify icons).

### Known Consumers

The Chromium bug has a significant amount of developer interest.

## Specification

### Schema

manifest.json
```
"icon_variants": [
  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png",
  },
  {
    "16": "dark16.png",
    "32": "dark32.png",
    "color_scheme": "dark",
  },
  {
    "16": "light16.png",
    "32": "light32.png",
    "color_scheme": ["dark", "light"]
  }
]
```

>   `icon_variants` requires an array with a set of icon groups objects. Each icon group consists of a set of icon paths and icon group matching criteria.

Reviewer 2: ```suggestion
  `icon_variants` requires an array with a set of icon groups objects. Each icon
group consists of a set of icon paths and icon group matching criteria.
```

>   The icon paths are set using the size as key and/or the keyword "any".

Reviewer 2: ```suggestion
  The icon paths are set using the size as key and/or the keyword `"any"`.
```

>   Optionally the `color_scheme` matching criterium which can either be an string or array of color schemes.
> Set icon_variants dynamically.

Reviewer 2: ```suggestion
  Optionally the `color_scheme` matching criterium which can either be an string or
array of color schemes.

Setting `icon_variants` dynamically:
```

```

A benefit of this new structure is that it's more resiliant to future changes,
thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
monochrome), and etc.

## setIcon()
Incumbent action.setIcon(), for reference.
```
action.setIcon({
  path,
  tabId
  imageData,
});
```

### Behavior

Existing manifest key. The behavior of the icons manifest key will remain
unchanged.
```
{
  "icons": {
    "64": "icon_64.png"
  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

The `dark` icon variant can be used by browsers to improve readability of the
icon. Often this will be when the OS or browser color scheme is dark (often
referred to as "dark mode") but also when the browser uses a darker shade theme
which has a background on which the dark icon variant would be more readable.
The `light` icon variant is used otherwise.

### New Permissions

N/A.

### Manifest File Changes

1. Any icon group that does not contain a "color_scheme" key will apply to all
available options, e.g. both "light" and "dark".

> 1. If "icon_variants" contains an icon group with matching conditions, the icon(s)

Reviewer 2: ```suggestion
1. If `icon_variants` contains an icon group with matching conditions, the icon(s)
```

specified in the first matching icon group based on insertion order will be used.
The other icon groups after that match will be ignored.

>    * "\<size\>" will be used used instead of `"any"` in case in case there are
>    matching conditions.

Reviewer 2: I'm not sure what this means? It is also covered in more detail below.
```suggestion
```

>    * If there is more than one matching icon object, any without `color_scheme` will be applied to every possible color scheme. Therefore, a subsequent matching icon object with a `color_scheme` will not be used.

Reviewer 2: ```suggestion
   * If there is more than one matching icon object, any without `color_scheme` will
be applied to every possible color scheme. Therefore, a subsequent matching icon
object with a `color_scheme` will not be used.
```

>    * Non-exact size matches will return an icon nearest the requested size, or the smallest in case of a tie.

Reviewer 2: This is already covered below.
```suggestion
```

   * Incorrectly typed `color-scheme` will be ignored, with an optional warning.

>    * If e.g. only one icon object is defined with a specific color scheme, that
>    icon object will be applied to all color schemes. It will be treated as the
>    default.

Reviewer 2: ```suggestion
   * If only one icon object is defined with a specific color scheme, that
   icon object will be applied to all color schemes. It will be the icon used.
```

1. If only one icon group is supplied, all matching conditions will be ignored.
1. icon_variants will not cause an error in the event that it is invalid
Instead, only the faulty icon group(s) will be ignored, with an optional
warning. Warnings are preferred over errors because they're more adaptable to
changes in the future.
1. Neither `"dark"` nor `"light"` color schemes are required.
1. If the top-level `icon_variants` key is provided, the top level `icons` key
will be ignored.

>    1. For `action` icons, the browser first checks for `icon_variants` in the `action`. If not specified, it falls back to `default_icon`, then to the top-level `icon_variants` or `icons`.
>    1. Any programmatically set icons using `action.setIcon()` would override the declaratively defined icons mentioned above.

Reviewer 2: ```suggestion
   1. For `action` icons, the browser first checks for `icon_variants` in the `action`.
If not specified, it falls back to `default_icon`, then to the top-level `icon_variants`
or `icons`.
   1. Any programmatically set icons using `action.setIcon()` would override the
declaratively defined icons mentioned above.
```

# Dark Mode Extension Icons

**Summary**

Feature to enable developers to enhance extension icon visibility in dark mode.

**Document Metadata**

**Author:** solomonkinard

> **Sponsoring Browser:** Chromium, Safari, Firefox

Reviewer 3: I think there was a thread about this, so apologies if I missed the resolution, but I think this should just be Chromium. This is not intended to represent the interested parties or even those that will implement, but whoever is specifically putting time into leading the proposal. No strong feelings on which vendor this is but I don't think having three makes sense. A good way of thinking about it is likely "which vendor should I contact for questions about the state of the proposal".

Author: Done.

  },
  {
    "16": "16.png",
    "32": "32.png",
  },
  {
    "16": "dark16.png",
    "32": "dark32.png",
    "color_scheme": "dark",
  },
  {
    "16": "light16.png",
    "32": "light32.png",
    "color_scheme": ["dark", "light"]
  }
]
```
  `icon_variants` requires an array with a set of icon groups objects. Each icon
group consists of a set of icon paths and icon group matching criteria.

  The icon paths are set using the size as key and/or the keyword `"any"`.

  Optionally the `color_scheme` matching criterium which can either be an string or
array of color schemes.

Setting `icon_variants` dynamically:
```
const exampleProperties: {string: string | ImageData}[] = [
  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png"
  },
  {
    "16": darkImageData16,
    "32": darkImageData32,
    "color_scheme": "dark"
  },
  {
    "16": lightImageData16,
    "32": lightImageData32,
    "color_scheme": "light"
  }
];

// action.setIcon().
const createProperties = {variants: exampleProperties};
action.setIcon(createProperties);

// menus.*(), for supporting browsers.
const menusProperties = {icon_variants: exampleProperties};
menus.create(menusProperties);

> menus.update(menusProperties);

Reviewer 6: ```suggestion
menus.update(id, menusProperties);
```

Note: menus is supported by Firefox and Safari; Chrome only supports contextMenus (but icons is not supported, yet? We just discussed this addition today). If you prefer, feel free to replace "icons" with "contextMenus" (which are also supported by Firefox and Safari as aliases of menus).

  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

The `dark` icon variant can be used by browsers to improve readability of the
icon. Often this will be when the OS or browser color scheme is dark (often
referred to as "dark mode") but also when the browser uses a darker shade theme
which has a background on which the dark icon variant would be more readable.
The `light` icon variant is used otherwise.

### New Permissions

N/A.

### Manifest File Changes

1. Any icon group that does not contain a "color_scheme" key will apply to all
available options, e.g. both "light" and "dark".

1. If `icon_variants` contains an icon group with matching conditions, the icon(s)
specified in the first matching icon group based on insertion order will be used.
The other icon groups after that match will be ignored.
   * If there is more than one matching icon object, any without `color_scheme` will
be applied to every possible color scheme. Therefore, a subsequent matching icon
object with a `color_scheme` will not be used.

>    * Incorrectly typed `color-scheme` will be ignored, with an optional warning.

Reviewer 3: ```suggestion
   * Incorrectly typed `color_scheme` values will be ignored, with an optional warning. For example, `color_scheme: ["unknown"]` will be treated as an empty array.
```

>    * Incorrectly typed `color_scheme` values will be ignored, with an optional warning. For example, `color_scheme: ["unknown"]` will be treated as an empty array.

Reviewer 2: The only remaining line that is not hard wrapped.

```suggestion
   * Incorrectly typed `color_scheme` values will be ignored, with an optional warning.
   For example, `color_scheme: ["unknown"]` will be treated as an empty array.
```

   * If only one icon object is defined with a specific color scheme, that
   icon object will be applied to all color schemes. It will be the icon used.

1. If only one icon group is supplied, all matching conditions will be ignored.

> 1. icon_variants will not cause an error in the event that it is invalid

Reviewer 6: Although this section is about the manifest, it also mentions setIcon.

Therefore I am asking the question here: can we be explicit about what happens when an unrecognized property (or unrecognized color scheme) is found in the API?

Chrome and Firefox have the convention in the extension API to throw from API methods, which enables feature detection at runtime.

Safari might be inclined to silently ignore unknown keys.

@xeenon thoughts?

Reviewer 2: `color_scheme` would be a known key, so we would throw for unknown values. We _would not_ throw for other keys (other than numbers or `color_scheme`).

Reviewer 4: @xeenon correct me if I am wrong, but I believe you mentioned unknown values in `color_scheme` should just be ignored by the browser? Why should an error be thrown in this situation?

Reviewer 2: @carlosjeurissen It would be ignored in the manifest, but I support it being an error in the JS API, since that allows feature detection in JS. We [currently throw](https://github.com/WebKit/WebKit/blob/main/Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm#L419) for invalid sizes or `ImageData` today for example, but these are silent in the manifest.

Reviewer 4: @xeenon It would be ideal if we can accommodate both the feature detection and the main benefit of this new syntax being the forgiving / future proof syntax. I believe we can achieve this by going for the following:

There are two kind of matching criteria. One being the property and one being values. When passing matching criteria property names in icon groups to `action.setIcon` which are not understood by the browser, the whole icon group will be skipped and marked as not matching. For property values, for properties taking multiple values (like `color_scheme`, the property values not understood by the browsers will silently be ignored. A property which ends up with no supported values will also be considered unsupported.

So this would mean `action.setIcon` would not throw an exception at all. Except, and here is where the feature detection comes in. If non of the icon groups have supported matching criteria. Instead of dropping all matching criteria as would be done for the manifest (I believe we agreed on this? might need some more specification), when no icon groups have supported matches, `action.setIcon` will throw an exception. 

This behaviour would mean feature detection can be done by passing one icon group. For example:
```js
browser.action.setIcon({
  variants: [{
    any: "path.svg",
    color_scheme: "dark"
  }]
}).catch((e) => {
  // throws if the browser does not support dark color scheme since no icon group is supported.
});
```

And still use the forgiving fallback nature of the syntax as such:
```js
browser.action.setIcon({
  variants: [{
    any: "path.svg",
    color_scheme: "dark"
  }, {
    16: "path.png"
}]
}).then(result => {
  // even tho the browser does not support dark color_scheme, it will not throw and simply ignore this icon group. 
});
````

This way we have the best of both worlds.

Reviewer 2: @carlosjeurissen I like that! Makes sense to me.

Author: A related "commit suggestion" was applied to the patch, which addresses this subthread:
```
`action.setIcon({variants})` will not throw when giving it icon groups with
properties it does not understand. Those icon groups will simply be ignored for
making matches. If however, none of the icon groups are supported, an exception
will be thrown allowing both feature detection and specifying fallbacks without
requiring feature detection.
```

  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

The `dark` icon variant can be used by browsers to improve readability of the
icon. Often this will be when the OS or browser color scheme is dark (often
referred to as "dark mode") but also when the browser uses a darker shade theme
which has a background on which the dark icon variant would be more readable.
The `light` icon variant is used otherwise.

### New Permissions

N/A.

### Manifest File Changes

1. Any icon group that does not contain a "color_scheme" key will apply to all
available options, e.g. both "light" and "dark".

1. If `icon_variants` contains an icon group with matching conditions, the icon(s)
specified in the first matching icon group based on insertion order will be used.
The other icon groups after that match will be ignored.
   * If there is more than one matching icon object, any without `color_scheme` will
be applied to every possible color scheme. Therefore, a subsequent matching icon
object with a `color_scheme` will not be used.
   * Incorrectly typed `color_scheme` values will be ignored, with an optional warning.

>    For example, `color_scheme: ["unknown"]` will be treated as an empty array.

Reviewer 4: ```suggestion
   For example, `color_scheme: ["unknown"]` will be treated as an empty array.
   An empty arrays means the browser should mark the icon group as invalid and ignore it.
```

will be ignored.
   1. For `action` icons, the browser first checks for `icon_variants` in the `action`.
If not specified, it falls back to `default_icon`, then to the top-level `icon_variants`
or `icons`.
   1. Any programmatically set icons using `action.setIcon()` would override the
declaratively defined icons mentioned above.
1. The `"16"` is a size in `{"16": "icon.png"}` and any number can be used as a size, per
the browser’s icon requirements. The word `"any"` can also be used in place of a number
to represent an icon that can be shown at any size (usually vector images). The icon size
used by the browser will be determined as follows:
   1. **Raster images:** Sizes are in pixels. For high-density devices (2x, 3x, etc.), the
browser looks for double or triple the desired point size (e.g., 32 or 48 pixels for a 16
point request). If the exact pixel size is unavailable, the next largest pixel size or `"any"`
will be used.
   1. **Vector images:** Sizes are in points, ensuring device independence. If the exact
point size is unavailable, an integer multiple (e.g. 32, 48, etc.) or `"any"` will be used.

> 

Reviewer 4: ```suggestion
If none of the specified icon groups have matching criteria, browsers should drop matching criterium in a specified order until it finds a group which results in a match. It will start by dropping any matching criterium which are unsupported/unknown. If still no match could be made, it will drop known matching criteria in a future agreed upon order.
```

Reviewer 2: Should add hard wraps so the document character width is consistent.

> 

Reviewer 4: ```suggestion
`action.setIcon({variants})` will not throw when giving it icon groups with properties it does not understand. Those icon groups will simply be ignored for making matches. If however none of the icon groups are supported, an exception will be thrown allowing both feature detection and specifying fallbacks without requiring feature detection.
```

Reviewer 2: Also hard wrap.

Author: This was included manually due to inability to click to commit some suggestions in the UI because of an "outdated" label.

theme. The set of dark mode icons can and will be made automatically visible
once the system setting is in dark mode, thereby enhancing the overall browser
experience.

#### Use Cases

1. Improved icon visibility in the extension toolbar.
1. Improved icon visibility on management the management and shortcuts pages.
1. Dark mode icon declarations made possible through the extension manifest.
1. setIcon() will allow setting of dark and/or light mode specific icons.
1. Improved icon visibility on context menu (Chrome and Safari reply on default
icons, only Firefox can specify icons).
1. Improved icon visibility on side panel (Chrome replies on default icons,
Firefox can specify icons).

### Known Consumers

The associated and linked bug has 48 stars and 31 comments.

## Specification

### Schema

Some browsers don't currently have plans to support svg. Therefore, a drop-in
replacement of an already supported image type will be allowed.

Shared IDL
```
// DOMString <mode> : "dark" | "light";

dictionary IconSizeToPath {
  DOMString <size> => DOMString <path>;
}

dictionary IconVariants {
  DOMString <mode> => IconSizeToPath;
}

> dictionary IconVariantsKey {

Reviewer 6: I'm trying to understand the type specification, but ran into a few hurdles. Here is a write-up of my thoughts, which hopefully highlights parts that can be clarified.

What is the purpose of this type? At first I was thinking of suggesting to rename "Key" to "Dict" because it should specify the value of the "icon_variants" object.

But then I saw that `action` also gets this key. The most reasonable interpretation of that is a "mixin" in IDL.

And then I tried to parse the value of this dictionary type and got somewhat stuck on trying to understand it. I think that you are using `<mode>` as a placeholder for light/dark defined above. When I try to apply that pattern here, I guess `<icon_variant>` is a placeholder for a number. But what does `<default_path>` mean?

Author: Unfortunately, icons aren't currently implemented in IDL. I decided to switch to IDL for this proposal to avoid inventing a new dialect. That said, can you provide a pointer -- my intent was to express what it it's supposed to be as DOMString isn't clear. It would be cleaner if there was a notion of `Path = DOMString` above, to then replace "DOMString <default_path>" here. The word default could be dropped as well. Any pointers on how to represent a DOMString with the aforementioned specificity in this proposal?

Reviewer 6: In Firefox, we don't use IDL either, but JSON schemas for the manifest.

But if you want to be closer to IDL, one simplification aspect is to recognize that property names are always strings. So `DOMString <size>` could be simplified to `<size>` with a description of what it could be.

An alternative could be to use Typescript syntax, which is explicitly mentioned as a way to specify the API: https://github.com/w3c/webextensions/blob/main/proposals/proposal_template.md#schema

I don't mind the exact syntax as long as it is obvious and unambiguous.

Reviewer 6: I found several examples of specifying manifest keys in idl in Chrome: https://cs.chromium.org/ManifestKeys%20file:extensions%20file:.idl

Author: tldr: TypeScript is used instead of IDL. ImageData is supported in action.setIcon() for icon_variants, and a future work section now includes exposing to developers whether the user preference is dark mode.

I decided to go for TypeScript instead. I also reinvented the inclusion of ImageData in setIcon for IconVariants. I think it's complex and yet easy to understand. However, implementation would also be complex. For that reason, I am not in favor of supporting ImageData in setIcon for anything related to icon_variants. However, it's included here in case no one else also objects to its inclusion, like me.

My argument against its inclusion is that setting ImageData is already a very advanced technique for developers and I think developers would be better serveed if they had an API endpoint to discover whether the user preference is dark mode or not. I'll add that new insight to the future work section.

Edit: On second thought, I could make a case not to expose the dark mode setting to developers. For that reason, I won't add it to the future work section at this time. That doesn't mean that I'm 100% behind ImageData support, it's just that I haven't thought of a better alternative so far. I don't like how complex ImageData is, how much work required to implement it, and no knowledge of how much it's even wanted by developers in the first place.

Reviewer: <img width="478" alt="Screenshot 2024-04-11 at 12 24 20" src="https://github.com/w3c/webextensions/assets/776467/a0a86062-6f76-4dc7-bb38-ee8cdff9234a">

On macOS, Activity Monitor can show mini CPU usage graphic on its dock icon. Using imageData to dynamically set extension icon is similar.

I have two extensions that set icon with imageData dynamically. They use (offscreen) canvas to draw a graphic and set it to icon. SetIcon is triggered by events or timer.

```
function generateIcon() {
    context2D.clearRect(0, 0, 32, 32);
    // draw some thing
    return context2D.getImageData(0, 0, 32, 32);
}

let imageData = generateIcon();
browser.action.setIcon({ imageData: { 32: imageData }});

// when supporting icon variants, the code might become:
let darkIcon = generateIcon('dark');
let lightIcon = generateIcon('light');
browser.action.setIcon(...);
```

Supporting icon variants in `action.setIcon()` is tricky and complex, because parameters `path` and `imageData` are orthogonal to the icon variation. So I suggest the following API shape:

```
action.setIcon({
    path,
    pathVariants,
    imageData,
    imageDataVariants,
    tabId
});

// for example
action.setIcon({
    pathVariants: {
        light: { 32: 'light-32.png' },
        dark: { 32: 'dark-32.png' },
    }
});

action.setIcon({
    imageDataVariants: {
        light: { 32: lightImageData },
        dark: { 32: darkImageData },
    }
});
```

Author: The `flat_option` seems to be getting the most attention as of most recently.

  DOMString <icon_variants> => IconVariants | DOMString <default_path>;
}
```

manifest.json IDL
```
dictionary ManifestKeys {
  IconVariantsKey? action;
  IconVariantsKey? icon_variants;
}
```

action.setIcon() IDL
```
interface SetIconArguments {
  ImageData | Object imageData;

>   DOMString | Object path?;

Reviewer 6: `| Object` are rather open types. Why?

(I think that they aren't part of this proposal. Feel free to add a comment to de-emphasize the importance of the exact type if wanted).

In another proposal, recently the "Multiple User Script Worlds" type from Devlin, I suggested the use of the diff syntax to show more clearly the new parts of a proposal in comparison to an existing one. If you use the same pattern here, then you wouldn't have to be concerned about the exact shape of the existing API format.

Author: > | Object are rather open types. Why?

Upgraded from an object.

> If you use the same pattern here, then you wouldn't have to be concerned about the exact shape of the existing API format.

Replaced incumbent parts with "...". It's earlier presence probably helped reviewers catch the fact that imageData was missing from `icon_variants`. That was intentional on my part, but it has now been included due to reviewer request.

    optional DoneCallback callback);
}
```

Action
```
action.setIcon({
  imageData?: ImageData | object:<numberToPathDictionary>,
  path?: string | object:<numberToPathDictionary>,
  tabId?: number,
  icon_variants?: {
    "<dark|light>": I
  },
});
```

### Examples

manifest.json
```
{
  "icon_variants": {
    "dark": {
      "128": "128_dark.png"
    },
    "light": {
      "128": "128_light.png"
    }
  },
  "action": {
    "default_icon": {
      "128": "128_action.png"
    },
    "icon_variants": {
      "dark": {
        "128": "128_action_dark.png"
      },
      "light": {
        "128": "128_action_light.png"
      }
    }
  }
}
```

Action
```
action.setIcon({
  variants: {
    "dark": {
      "128": "128_action_dark.png"
    },
    "light": {
      "128": "128_action_light.png"
    }
  },
});
```

### Behavior

Existing manifest key. The behavior of the icons manifest key will remain unchanged.
```
{
  "icons": {
    "64": "icon_64.png"
  },
  "action": {
    "default_icon": {
      "64": "action_64.png"
    }
  }
}
```

Order of precedence. The new `icon_variants` keys and subkeys will take
precedent, followed by the incumbent `icons` key.

> 

Reviewer 6: To avoid any ambiguity, let's explicitly call out the intended semantic of dark/light.

```suggestion

The `dark` icon variant is used when a dark color scheme is used for the
surrounding UI, often referred to as "dark mode".
The `light` icon variant is used otherwise.

```

#### Fallback
* `icon_variants` will be used if they're supplied. Otherwise `default_icon`
will be used if it's supplied. If neither are supplied, then `icons` will be
used.

* icon_variants or variants in action.setIcon() must provide both dark and light
icon, if they occur.

* Order of precedence for action:
<api>.action.setIcon() -> action.icon_variants -> action.default_icon ->
icon_variants -> icons

* Order of precedence for non-action:
icon_variants -> icons

### FAQ
* What happens if you specify icon_variants.dark but not icon_variants.light,
and also icons?
If either is provided, both are expected.

Do we use the dark mode icon from icon_variants for light as well, or fallback
to icons?
Both are expected if either "light" or "dark" are supplied.

What happens if you specify icon_variants.dark but not icon_variants.light, and
you don't have icons?
This could be a warning and a silent fallback to "icons", as the goal is to try
to reduce errors as much as possible.

Do we use the dark mode icon from icon_variants or show an auto-generated icon?
The dark mode icon from variants will be shown if supplied and the user is
currently in dark mode.

#### Warning or error?

> The new `icon_variants` key (or sub-key) will only emit a non-blocking warning

Reviewer 6: Can you rephrase this? The current proposal makes "warning" a requirement, while the actual intent is to not have an error. Emitting neither would also be acceptable.

In Firefox we would probably emit a warning for invalid manifest keys.

Author: Done.

theme.

#### Use Cases

1. Improved icon visibility in the extension toolbar.
1. Improved icon visibility on the management and shortcuts pages.
1. Dark mode icon declarations made possible through the extension manifest.
1. `setIcon()` will allow setting of dark and/or light mode specific icons.
1. Improved icon visibility on context menu (Chrome relies on default
icons, only Firefox and Safari can specify menu icons).
1. Improved icon visibility on side panel (Chrome relies on default icons,
Firefox can specify icons).

### Known Consumers

The Chromium bug has a significant amount of developer interest.

## Specification

### Schema

manifest.json
```
"icon_variants": [
  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png",
  },
  {
    "16": "dark16.png",
    "32": "dark32.png",
    "color_scheme": "dark",
  },
  {
    "16": "light16.png",
    "32": "light32.png",
    "color_scheme": ["dark", "light"]
  }
]
```
  `icon_variants` requires an array with a set of icon groups objects. Each icon
group consists of a set of icon paths and icon group matching criteria.

  The icon paths are set using the size as key and/or the keyword `"any"`.

>   Optionally the `color_scheme` matching criterium which can either be an string or

Reviewer 2: ```suggestion
  Optionally the `color_scheme` matching criterium which can either be a string or
```

array of color schemes.

Setting `icon_variants` dynamically:
```
const exampleProperties: {string: string | ImageData}[] = [
  {
    "any": "any.svg",
  },
  {
    "16": "16.png",
    "32": "32.png"
  },
  {
    "16": darkImageData16,
    "32": darkImageData32,
    "color_scheme": "dark"
  },
  {
    "16": lightImageData16,
    "32": lightImageData32,
    "color_scheme": "light"
  }
];

// action.setIcon().
const createProperties = {variants: exampleProperties};
action.setIcon(createProperties);

// menus.*(), for supporting browsers.
const menusProperties = {icon_variants: exampleProperties};
menus.create(menusProperties);
menus.update(id, menusProperties);
```

A benefit of this new structure is that it's more resiliant to future changes,
thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
monochrome), and etc.

## setIcon()

> Incumbent action.setIcon(), for reference.

Reviewer 2: ```suggestion
Incumbent `action.setIcon()`, for reference.
```

precedent, followed by the incumbent `icons` key.

The `dark` icon variant can be used by browsers to improve readability of the
icon. Often this will be when the OS or browser color scheme is dark (often
referred to as "dark mode") but also when the browser uses a darker shade theme
which has a background on which the dark icon variant would be more readable.
The `light` icon variant is used otherwise.

### New Permissions

N/A.

### Manifest File Changes [And API]

**Summary**

`icon_variants` is offered as a replacement for `icons`. `icon_variants` aims
to be more flexible than pre-existing keys, allowing for unexpected declarations
without causing errors that prevents extension installation. Absent
`color_scheme` declaration, any matching individual `icon_variants` group will
effectively have a wild-card `color_scheme`.

**Misc**
1. If the top-level `icon_variants` key is provided, the top level `icons` key
will be ignored.
1. `icon_variants` will not cause an error in the event that it is invalid.
Instead, only the faulty icon group(s) will be ignored, with an optional
warning. Warnings are preferred over errors because they're more adaptable to
changes in the future.

> 1. `color-scheme`. Any icon group that does not contain a "`color_scheme`" key

Reviewer 2: ```suggestion
1. `color-scheme`. Any icon group that does not contain a `color_scheme` key
```

**Created:** 2024-04-05

**Related Issues:**
* https://crbug.com/893175
* https://github.com/w3c/webextensions/issues/229

## Motivation

### Objective

Extension developers will be able to supply and define a set of icons to be used
in the event that the user has expressed the preference for a page that has a
dark theme.

#### Use Cases

1. Improved icon visibility in the extension toolbar.
1. Improved icon visibility on the management and shortcuts pages.
1. Dark mode icon declarations made possible through the extension manifest.
1. `setIcon()` will allow setting of dark and/or light mode specific icons.
1. Improved icon visibility on context menu (Chrome relies on default
icons, only Firefox and Safari can specify menu icons).
1. Improved icon visibility on side panel (Chrome relies on default icons,
Firefox can specify icons).

### Known Consumers

The Chromium bug has a significant amount of developer interest.

## Specification

### Schema

manifest.json
```
type ImagePath = string;
type ColorScheme = "light" | "dark";

interface IconVariantsBase {
  color_scheme?: ColorScheme | ColorScheme[];
}

interface IconVariantsManifest extends IconVariantsBase {

>   any?: ImagePath;

Reviewer 3: I think this should be something like `[key: string]`? What we have now is a literal key with the name "any".

  [index: number]: ImagePath;
}

interface ManifestAction {

>   [icon_variants: string]: IconVariantsManifest;

Reviewer 3: @dotproto, shouldn't this just be `icon_variants: IconVariantsManifest[]`? The current syntax is saying that **any** key added under "action." needs to be a single IconVariantsManifest and will be given the name `icon_variants` for typing reasons.

Author: (resolving all comments, as per meeting, as all merge-blocking comments should have been addressed.)

}

interface ExtensionManifest {
  icon_variants: IconVariantsManifest[];
  action: Action;
}
```

action.setIcon
```
interface SetIconVariantsApiDetails extends IconVariantsBase {
  any?: ImagePath | ImageData;
  [index: number]: ImagePath | ImageData;
}

interface SetIconVariantsApi {

>   [variants: string]: ActionIconVariants;

Reviewer 3: Similar to above, I'm not sure these types do what we expect.

# Dark Mode Extension Icons

**Summary**

Feature to enable developers to enhance extension icon visibility in dark mode.

**Document Metadata**

**Author:** &lt;solomonkinard>

**Sponsoring Browser:** Chromium.

**Contributors:** &lt;oliverdunk>

**Created:** 2024-04-05

**Related Issues:**
* https://crbug.com/893175
* https://github.com/w3c/webextensions/issues/229

## Motivation

### Objective

Extension developers will be able to supply and define a set of icons to be used
in the event that user has expressed the preference for a page that has a dark
theme. The set of dark mode icons can and will be made automatically visible
once the system setting is in dark mode, thereby enhancing the overall browser
experience.

#### Use Cases

1. Improved icon visibility in the extension toolbar.
1. Improved icon visibility on management the management and shortcuts pages.
1. Dark mode icon declarations made possible through the extension manifest.
1. setIcon() will allow setting of dark and/or light mode specific icons.

> 1. Improved icon visibility on context menu (Chrome and Safari reply on default

Reviewer 2: Safari Technology Preview 192 does support `menus.create()` and `menus.update()` with the `icons` key like Firefox. We should support the same `icon_variants` key in the `menus` API as `action.setIcon()`.

Author: Done.

icons, only Firefox can specify icons).
1. Improved icon visibility on side panel (Chrome replies on default icons,
Firefox can specify icons).

### Known Consumers

The Chromium bug has a significant amount of developer interest.

## Specification

### Schema

// TODO: Obtain consensus on any of the following options before merging.

Incumbent action.setIcon(), for reference.
```
action.setIcon({
  path,
  tabId
  imageData,
});
```

---

#### Option (donut option)

Don't implement setIcon() for dark mode, as it's used for dynamic icons anyway.

* Pros: Easy to implement.

> * Cons: Developers would need to find their own ways of detecting dark mode.

Reviewer: If this proposal doesn't support `action.setIcon()` for dark and light icons. Can browsers provide APIs for detecting the browser (not OS) dark/light theme and a onChangeListener? For example:

```
browser.runtime.getBrowserTheme() // return 'dark' or 'light'

browser.runtime.onBrowserThemeChanged.addListener((newTheme) => {
    // newTheme is dark or light
    action.setIcon(...);
});
```

Reviewer: In addition, Firefox and Safari support setting context-menu item icon (See #592), and there is a proposal (#550) for setting sidebar icon. So besides `action.setIcon()` for the toolbar icon, there are other places where developers might need to dynamically set icons for dark/light theme in the future. So, in my option, supporting `runtime.getBrowserTheme()` and `runtime.onBrowserThemeChanged` might be a more general and easy-to-implement solution for dynamical icons.

Reviewer 2: I do not support an event for this. The browser needs the color scheme icons before the user switches modes, otherwise there will be stale icons as we wait for the extension to respond to the event.

`menus.create()` and `menus.update()` should just support light/dark mode variants too.

Reviewer: > I do not support a change event for this. The browser needs the color scheme icons before the user switches modes, otherwise there will be stale icons as we wait for the extension to respond to the event.

This is an important point that affects the design of the API. I'd like to know other browsers' opinions.

Author: I think there have been sufficient ideas shared to come to an understanding, represented in the latest patch.

Reviewer: For me, as a developer, I don't think the change event listener or delaying some milliseconds is a problem. Instead, the other way requires developers to generate two icons every time and the browser need to hold two icons in memory, but only use one of them (approximately 4 KB of memory is wasted).

Anyway, I don't mind how it's implemented for setIcon API, as long as the functionality is there.

Reviewer 4: @hanguokai Also maintaining extensions which use `imageData` for `setIcon`. Especially with different densities and the lack of figuring out what the current dark mode and devicePixelRatio is in serviceWorkers. The amount of variants could add up especially when different tabs are involved.

If we would provide this info to extension developers, I believe it should be on the action namespace and include more data. Something like:

`browser.action.getIconVariables()` which would give an object with data about the current conditions:
```json
{
  "color_scheme": "dark",
  "size": 24
}
```

And optionally
`browser.action.onIconVariablesChanged.addListener((newIconVariables) => { });`

Knowing the size the action icon will be used seems like a much bigger win than knowing the `color_scheme` as it is more stable and will provide more memory gains and more crispy icons when `imageData` is being used.

action.setIcon()
```
action.setIcon({
  variants: {
    "dark": {
      "128": "128_action_dark.png"
    },
    "light": {
      "128": LightImageData
    }
  },
});
```

Schema

manifest.json
```
// `Mode` is dark or light (if either is used, both are expected).
const Modes: string[] = ["dark", "light];
const Mode = Modes[Math.floor(Math.random() * Modes.length())];

// Primitive type declaration.
type Size = string;
type Path = string;

// Map icon size to path.
const IconSizeToPath: map<Size, Path>;

// Map mode to a string path or a dictionary mapping icon size to a string path.
const IconVariants: map<Mode, IconSizeToPath | Path>;

{
  ...
  "icon_variants": IconVariants;
  "action": {
    ...
    "icon_variants": IconVariants
  },
};
```

action.setIcon()
```
// ImageData is an interface representing canvas element pixels.
const IconSizeToImageData: map<Size, Path>;
const IconVariantsWithImageData:
  map<Mode, IconSizeToPath | IconSizeToImageData | ImageData | Path>;
const ImageDataDictionary: map<Size, ImageData>;

action.setIcon(
  ...
  icon_variants?: IconVariantsWithImageData,
);
```

* Pros: Easy for developers to use.
* Cons: Complex to define and implement.

---

> #### Option (flat option)

Reviewer 2: This is the option I prefer. If we do this one, unknown keys should be non-fatal, so we can expand this in the future without too much issues with backwards compatibility or the need to abandon this for a new key because browsers were too fussy with errors. (Like if we did add a `type` sub-key down the line.)

Reviewer 3: This is also my preferred approach.

Should we say that you can specify an icon for the same theme twice and the browser will use either the first or last one it can load? For example, it would be nice to support:

```
"icon_variants": [
  {
    "16": "16.svg"
  },
  {
    "16": "16.png"
  },
  {
    "16": "dark16.svg"
    "color_scheme": "dark",
  },
  {
    "16": "dark16.png"
    "color_scheme": "dark",
  }]
```

With the understanding browsers that support .svg would use that icon, and those that don't would fallback to the png.

An alternative would be allowing the key to take an array of possible icon files.

Reviewer 2: @oliverdunk Sounds good to me. I like having first usually icon win here.

I'm less keen on having the path be an array, but I wouldn't oppose it if others prefer that.

Author: That sounds ok: path as a string, not an array.

Reviewer 4: If we are going down the path of first icon win (which seems to be preferred). This would mean browsers should ignore icon objects/groups (We should probably find a proper name for this) which have keys it does not understand. Correct me if I am wrong.

manifest.json
```
"icon_variants": [{
  "16": "16.png",
  "32": "32.png",
}, {
  "16": "dark16.png",
  "32": "dark32.png",
  "color_scheme": "dark",
}, {
  "16": "light16.png",
  "32": "light32.png",
  "color_scheme": "light"
}]

> ```

Reviewer 2: I think this option should allow `light` to be assumed for the case that is missing `color_scheme`.

```json
[{
  "16": "16.png",
  "32": "32.png",
}, {
  "16": "dark16.png",
  "32": "dark32.png",
  "color_scheme": "dark",
}]
```

Also this option should allow `"color_scheme": ["light", "dark"]` like the CSS property, so a universal icon that works in both modes can be set once. This lets the developer use the new property / methods, and not need to fallback to the old `icons` or `default_icon` keys.

```json
[{
  "16": "16.png",
  "32": "32.png",
  "color_scheme": ["light", "dark"]
}]
```

One final note, we should allow an `"any"` size in this new proposal, so SVG is a first class citizen. At least carve it out so `"any"` is not an error and can be supported by browsers that support SVG icons. (This matches what Web App Manifest does.)

```json
[{
  "any": "icon.svg",
  "color_scheme": ["light", "dark"]
}]
```

Reviewer: In this flatten option, if we follow [PWA manifest format](https://developer.mozilla.org/en-US/docs/Web/Manifest/icons), it should use `sizes` as the key, rather than "16", "32", "any" as keys.
```
[
  {
    "src": "icon/48.png",
    "sizes": "48x48"
  },
  {
    "src": "icon/icon.svg",
    "sizes": "any"
  }
]
```

I like your "color_scheme" suggestion.

Reviewer 2: I don't think we need to completely follow the Web App Manifest format here. I just wanted `any` to be considered. The web app format is too verbose in my opinion, and does a poor job of grouping similar icons. Their syntax is mostly severed by the need to support varying image formats, where our need is mostly about varying color schemes. We also have the benefit of using file extensions for type info, where they needed a `type` key.

Reviewer 3: @xeenon, is there a reason you would like to support only defining a light icon? If not, could we make the default `"color_scheme": ["light", "dark"]` instead of just light? That would allow you to omit the key in the case where the icon works for both and we wouldn't need to support an array there.

Reviewer 2: @oliverdunk That works too! I like that better. (We should still allow setting this as an array explicitly too.)

Author: `any` has been included. So has ["dark", "light"].

Reviewer 4: @hanguokai In the past during one of the meetings I proposed to use the same syntax as the WebManifest. This was considered too verbose. Which I have to agree with. Seems the syntax used by the WebManifest is designed to match `<link rel="icon">`. This is why I proposed the hybrid between what we had in webExtensions and webManifests. As many of the descriptors (`color_scheme`, `type`)  applies to a group of icons.

Reviewer 4: @xeenon Have to agree with @oliverdunk on making `"color_scheme": ["light", "dark"]` the default. Historically extension icons have suggested by at least Google Chrome to be designed to work for both `light` and `dark` extension themes. @xeenon What is the use-case for providing an array instead of simply omitting it? For additional code clarity or do you expect additional color_schemes to be added in the future?

Reviewer 2: @carlosjeurissen Yeah, array for future proofing really, and it matches the `color-scheme` CSS property supporting multiple values.

Reviewer 4: @xeenon Say additional items will be added to "color_scheme" in the future (which I do not expect), how would the browser handle this? Should it simply ignore schemes it does not understand?

Still not convinced why having `color_scheme` set to `["light", "dark"]` has any value. I feel like omitting it would say "do not filter on color_scheme".

If it happens to become useful in the future. We can always add it later. Restricting a specific icon to only be used for "light" and "dark" but not any other `color_scheme` if added does not seem like a use case.

Reviewer 2: @carlosjeurissen Yes, browsers should indeed ignore unknown color schemes. As long as the array includes at least one supported scheme, it should be valid. Unsupported `color-scheme` entries can simply be skipped during matching.

While designing the `color-scheme` CSS property (originally called [`supported-color-schemes`](https://github.com/WebKit/WebKit/commit/863d1a65f7df95314b2e0bdf1763d0c5d133fb07)), I considered incorporating high-contrast schemes, which are supported by macOS, Windows, and iOS. This could be a potential area for future enhancements.

It's crucial to anticipate possible future enhancements to prevent creating restrictions that might inhibit those future capabilities. Restricting ourselves initially to string values without supporting arrays would permanently confine us to a single value. This rationale is similar to why `preferred_environment` accepts both strings and arrays.

Allowing developers to supply schemes like `["light", "dark"]` does not prevent support for additional schemes in the future. However, if only an array of color schemes is supplied, any new schemes such as `dark-high-contrast` might necessitate fallback to the nearest best scheme, such as `dark`, which is the approach for apps on macOS and iOS. Not supplying a `color-scheme` implies "this icon will work for any color scheme now and in the future" — so there is a difference between an array of all currently defined color schemes and no value.

Allowing developers to supply all _supported color schemes_ for their icons gives them the power to be forward-looking and cautious.

Reviewer 4: @xeenon Convinced. Thanks for the detailed explanation. All in favor for future proofing ourselves. We should attempt to not require manifest version upgrades as much as possible.

action.setIcon()
```
action.setIcon({
  variants: [
    {
      "16": "16.png",
      "32": "32.png"
    }, {
      "16": darkImageData16,
      "32": darkImageData32,
      "color_scheme": "dark"
    }, {
      "16": lightImageData16,
      "32": lightImageData32,
      "color_scheme": "light"
    }
  ]
});
```

* Pros: Future resistant allowing for more keys such as density (e.g. 2dppx),
purpose (e.g. monochrome), and etc.
* Cons: New paradigm for icon definition, and unclear how to handle color_schemes.

> * Questions: Should `color_scheme` be used instead of `mode`?

Reviewer 4: The term `color_scheme` is indeed preferred over `mode` as it is used in other places on the web. For example, in the [`color-scheme` css property](https://developer.mozilla.org/en-US/docs/Web/CSS/color-scheme) and in the [`prefers-color-scheme` media feature](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme).

Author: Done.

`icon_variants` is an optional key of the top level manifest dictionary.
`icon_variants` is also an optional sub-key in the action key dictionary.

#### Fallback
* `icon_variants` will be used if they're supplied. Otherwise `default_icon`
will be used if it's supplied. If neither are supplied, then `icons` will be
used.

* icon_variants or variants in action.setIcon() must provide both dark and light
icon, if they occur.

* Order of precedence for action:
<api>.action.setIcon() -> action.icon_variants -> action.default_icon ->
icon_variants -> icons

* Order of precedence for non-action:
icon_variants -> icons

### FAQ
* What happens if you specify icon_variants.dark but not icon_variants.light,
and also icons?

If either is provided, both are expected.

* Do we use the dark mode icon from icon_variants for light as well, or fallback
to icons?

Both are expected if either "light" or "dark" are supplied.

* What happens if you specify icon_variants.dark but not icon_variants.light, and
you don't have icons?

This could be a warning and a silent fallback to "icons", as the goal is to try
to reduce errors as much as possible. Another option could be to fall back to

> the dark mode icon if in light mode and no light mode icon is supplied.

Reviewer 2: All three of these seem to be the asking / answering the same thing slightly differently.

Reviewer 3: This was my impression as well. A few thoughts:

- We should definitely support specifying light mode and not dark mode. Not all developers care about dark mode and having to supply both would be a burden.
- If you specify dark mode and not light mode, I have no strong feelings on if that is an error or not. I think "you must at least specify light mode" is a fair requirement.
- Falling back to `icons` from `icon_variants` seems nice in some ways but also introduces complexity, so I'm leaning towards saying we don't do that, but no strong feelings.

Reviewer: If the developer only provides one icon, whether it's dark or light , it doesn't make much sense to use the new "icon_variants". In this case, they can use current "icon" and "action.default_icon" which already means both dark and light icon. Therefore, the "icon_variants" is optional, if it is not provided, then fallback to others. But once provided, it is considered to represent all variant icons (i.e. if only provides one icon, it means both).

Reviewer 2: @hanguokai `icon_variants` is quickly becoming a way to [specify alternate icon formats too](https://github.com/w3c/webextensions/pull/585#discussion_r1567170430), so having it only support one `color_scheme` for those extensions that don't care (or have universal icons that work anywhere) seems best.

I see `icon_variants` as a fixed version of `icons` and `default_icon` and it should be a viable option for any developer that wants to use the new thing.

Reviewer: @xeenon A point of my [last comment](https://github.com/w3c/webextensions/pull/585#discussion_r1567403822) is that "once "icon_variants" is provided, it is considered to represent all variant icons (i.e. if only provides one icon, it means both).". What I mean is, if the user only provides one icon (`color_scheme` is dark or light, or no `color_scheme`), this icon is used for both dark and light icon. It doesn't fallback to "icon" or "action.default_icon".

Reviewer: > I see icon_variants as a fixed version of icons and default_icon and it should be a viable option for any developer that wants to use the new thing.

Oh, I understand, you mean "icon_variants" is not just a supplement to "icons" or "default_icon", "icon_variants" can completely replace them. If developers provide "icon_variants", they don't need to provide "icons" anymore.

Author: Synthesized these ideas as a replacement to the FAQ.

Reviewer 4: Also leaning towards regarding `icon_variants` as a replacement for `icons` in the long run. just like `options_ui` has mostly replaced `options_page`. This will motivate extension developers to add more `icon_variants` as well for a better user experience.

* Do we use the dark mode icon from icon_variants or show an auto-generated icon?

The dark mode icon from variants will be shown if supplied and the user is
currently in dark mode.

#### Warning or error?
The new `icon_variants` key (or sub-key) will not cause an error in the event
that it is invalid. Instead, either ignore it altogether, or emit a warning.
Warnings are preferred over errors as they're future-resilient.

## Security and Privacy

### Exposed Sensitive Data

N/A.

### Abuse Mitigations

N/A.

### Additional Security Considerations

N/A.

## Alternatives

### Existing Workarounds

1. A developer could ask the user to specify what mode they're in and then
dynamically set the icon to a dark mode icon using action.setIcon(). That
wouldn't change the management page icon.
2. A developer could change icons dynamically to dark mode icons if this is

> true: `window.matchMedia('(prefers-color-scheme: dark)')`.

Reviewer: This proposal focuses on matching the Dark/Light theme of the **Browser UI** (not OS UI). Note that the browser theme can be set independently of the operating system theme.

At present, the behavior of `window.matchMedia('(prefers-color-scheme: dark)').matches` is not consistent across browsers.
- Safari doesn't provide browser dark/light theme setting, it follows OS's theme.
- On Firefox, `prefers-color-scheme: dark` follows the browser setting, not the OS setting. IMO, this is the correct behavior.
- On Chrome, the behavior is very weird. See the following table.

| OS Theme | Chrome Theme | Web pages and Extension pages |
| ------------- | ------------- | ------------- |
| Light | Light | false (Light) |
| Light | Dark | false (Light) |
| Dark | Light | true (Dark) |
| Dark | Dark | true (Dark) |

On web pages or extension pages in a Chrome tab, `prefers-color-scheme: dark` follows the OS setting, not the browser setting.

| OS Theme | Chrome Theme | A override new-tab extension page |
| ------------- | ------------- | ------------- |
| Light | Light | false (Light) |
| Light | Dark | still false (Light), but refresh the page, it become true (Dark) |
| Dark | Light | still true (Dark), but refresh the page, it become false (Light)  |
| Dark | Dark | true (Dark) |

On a Chrome New-Tab override extension page, `prefers-color-scheme: dark` follows the Browser setting, not the OS setting, but you need to refresh the page to get the updated browser theme when changing the browser theme.

IMO, on Chrome, the behavior of `prefers-color-scheme: dark` is not correct.

Reviewer 2: I agree this sounds like a bug in Chrome with `prefers-color-scheme`, but it shouldn't impact this proposal. I expect extensions to follow the browser theme for the purposes of this proposal.

Author: I could either remove the workarounds or list even more possibilities. However, the workaround section is just meant to give an idea of possible mitigations while waiting for full dark-mode support.

Reviewer 4: @hanguokai Thanks so much for putting this together! Sounds like very odd behaviour in Chrome. Is there already a crbug on this one?

As for this proposal. I believe we should allow browsers to de-couple `prefers-color-scheme: dark` from using `color_scheme: dark` for extension icons. This allows browsers to have light-theme mode in general, while using the darker variant for extension icons when it uses a browser theme which is more dark than light (for example, green like the comment button on GitHub).

Reviewer: @carlosjeurissen I reported an issue [334682520](https://issues.chromium.org/issues/334682520) last week which marked as duplication of issue [332328864](https://issues.chromium.org/issues/332328864). Both issues are relatively new and not fixed. I think this is a Chrome internal issue. From a user's perspective, it's pretty obvious.

From a practical point of view, `prefers-color-scheme: dark` as a workaround is not ideal (such as theme changing problem). And that is exactly what this proposal seeks to address.

Reviewer 4: @hanguokai Not sure it hurts if browsers decide to use a dark-shaded theme while using light mode for web content. Some users actually enjoy this mode. The only odd behaviour seems to be the refreshing on the new tab page.

> From a practical point of view, prefers-color-scheme: dark as a workaround is not ideal 

That is exactly my point. Very much in favour of the `color_scheme` feature in this proposal.

Reviewer: > Not sure it hurts if browsers decide to use a dark-shaded theme while using light mode for web content. Some users actually enjoy this mode.

If so, the browser should provide two settings for them separately (not one setting). But that's too complicated, and I don't think browsers will do that.

Reviewer 4: > If so, the browser should provide two settings for them separately (not one setting). But that's too complicated, and I don't think browsers will do that.

This seems to be the status quo. If you select a dark theme from the chrome web store, and your OS is light mode. Web pages will still display in light mode. The same when selecting a light theme from the chrome web store while your OS is in dark mode. It will show webpages in dark mode.

Reviewer: @carlosjeurissen I didn't consider extension-themes before, only the (browser built-in) mode setting. Without installing a theme, the mode should change both UI and web pages. Without installing a theme, if the browser want users custom them separately, the browser should provide two settings, one setting for UI and one setting for web pages.

Back to the topic, this proposal is to solve the matching problem between the extension icon and the browser UI. `prefers-color-scheme` is a CSS function, that used for web pages. And @solomonkinard said "the workaround section is just meant to give an idea of possible mitigations while waiting for full dark-mode support.". So it doesn't really matter whether to delete this workaround in the proposal.

Reviewer 3: > And @solomonkinard said "the workaround section is just meant to give an idea of possible mitigations while waiting for full dark-mode support.". So it doesn't really matter whether to delete this workaround in the proposal.

Right, I would say there are two purposes to this section:

- Share possible workarounds to interested developers.
- Give context for what we are trying to make easier (if the workarounds were perfect, maybe we wouldn't need this proposal).

I don't think either of those meaningfully change us wanting to do this, so agreed with Jackie that I don't think we need to worry too much about this section :)

Reviewer 4: @oliverdunk Agreed there is no need to make updates to the section. Was just following up on the conversation about how browsers handle theming and dark mode.

**Document Metadata**

**Author:** solomonkinard

**Sponsoring Browser:** Chromium, Safari, Firefox

> **Contributors:** oliverdunk, xeenon, carlosjeurissen

Reviewer: Suggest adding myself (hanguokai) as a contributor since I have been actively promoting and participating in this issue for years.

Author: Done.

### Schema

manifest.json
```
type ImagePath = string;
type ColorScheme = "light" | "dark";

interface IconVariantsBase {
  color_scheme?: ColorScheme | ColorScheme[];
}

interface IconVariantsManifest extends IconVariantsBase {
  any?: ImagePath;
  [index: number]: ImagePath;
}

interface ManifestAction {
  [icon_variants: string]: IconVariantsManifest;
}

interface ExtensionManifest {
  icon_variants: IconVariantsManifest[];

>   action: Action;

Reviewer 3: I think this was the intention? `Action` isn't a type which exists elsewhere.

```suggestion
  action: ManifestAction;
```

manifest.json
```
"icon_variants": [
  {
    "any": "any.svg",

>   },

Reviewer 2: Maybe include an example of `"color_scheme": ["light", "dark"]` here, since you have the example below that has no `color_scheme` too, then all cases are covered.

Author: Done.
